cli/cmd/logs.go

```diff
@@ -0,0 +1,224 @@
+		config.LabelSelector = labels.Everything()
+	} else {
+		var podExists string
+		for _, p := range controlPlaneComponents {
+			if p == o.controlPlaneComponent {
+				podExists = p
+				break
+			}
+		}
+
+		if podExists == "" {
+			return nil, fmt.Errorf("control plane component [%s] does not exist. Must be one of %v", o.controlPlaneComponent, controlPlaneComponents)
+		}
+		selector, err := labels.Parse(fmt.Sprintf("linkerd.io/control-plane-component=%s", o.controlPlaneComponent))
+		if err != nil {
+			return nil, err
+		}
+		config.LabelSelector = selector
+	}
+
+	if o.container != "" {
+		var matchingContainer string
+		for _, c := range availableContainers {
+			if o.container == c {
+				matchingContainer = c
+				break
+			}
+		}
+		if matchingContainer == "" {
+			return nil, fmt.Errorf("container [%s] does not exist in control plane [%s]", o.container, controlPlaneNamespace)
+		}
+	}
+
+	containerFilterRgx, err := regexp.Compile(o.container)
+	if err != nil {
+		return nil, err
+	}
+	config.ContainerQuery = containerFilterRgx
+
+	if o.tail != -1 {
+		config.TailLines = &o.tail
+	}
+
+	// Do not use regex to filter pods. Instead, we provide the list of all control plane components and use
+	// the label selector to filter logs.
+	podFilterRgx, err := regexp.Compile("")
+	if err != nil {
+		return nil, err
+	}
+	config.PodQuery = podFilterRgx
+	config.Since = o.sinceSeconds
+	config.Timestamps = o.timestamps
+	config.Namespace = controlPlaneNamespace
+
+	return config, nil
+}
+
+func getControlPlaneComponentsAndContainers(pods *v1.PodList) ([]string, []string) {
+	var controlPlaneComponents, containers []string
+	for _, pod := range pods.Items {
+		controlPlaneComponents = append(controlPlaneComponents, pod.Labels["linkerd.io/control-plane-component"])
+		for _, container := range pod.Spec.Containers {
+			containers = append(containers, container.Name)
+		}
+	}
+	return controlPlaneComponents, containers
+}
+
+func newLogCmdConfig(options *logsOptions, kubeconfigPath, kubeContext string) (*logCmdConfig, error) {
+	// Check that we can call the Kubernetes API
+	_ := cliPublicAPIClient()
+	kubeAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext)
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(kubeAPI.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	podList, err := clientset.CoreV1().Pods(controlPlaneNamespace).List(meta_v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	components, containers := getControlPlaneComponentsAndContainers(podList)
+
+	c, err := options.toSternConfig(components, containers)
+	if err != nil {
+		return nil, err
+	}
+
+	return &logCmdConfig{
+		clientset,
+		c,
+	}, nil
+}
+
+func newCmdLogs() *cobra.Command {
+	options := newLogsOptions()
+
+	cmd := &cobra.Command{
+		Use:   "logs [flags]",
+		Short: "Tail logs from containers in the linkerd control plane",
+		Long:  `Tail logs from containers in the linkerd control plane`,
```

Comment:
For consistency with the help text from other commands (e.g. `dashboard`), we should use capital-L "Linkerd" in both the short and long descriptions here.